workflow/controller/controller.go

```diff
@@ -259,27 +259,37 @@ func (wfc *WorkflowController) periodicWorkflowGarbageCollector(stopCh <-chan st
 		case <-ticker.C:
 			if wfc.offloadNodeStatusRepo.IsEnabled() {
 				log.Info("Performing periodic workflow GC")
-				oldUIDs, err := wfc.offloadNodeStatusRepo.ListOldUIDs(wfc.GetManagedNamespace())
+				oldRecords, err := wfc.offloadNodeStatusRepo.ListOldOffloads(wfc.GetManagedNamespace())
 				if err != nil {
-					log.WithField("err", err).Error("Failed to list offloaded nodes")
+					log.WithField("err", err).Error("Failed to list old offloaded nodes")
 				}
-				if len(oldUIDs) == 0 {
-					log.Info("Zero old UIDs, nothing to do")
+				if len(oldRecords) == 0 {
+					log.Info("Zero old records, nothing to do")
 					return
 				}
-				list, err := util.NewWorkflowLister(wfc.wfInformer).List()
-				if err != nil {
-					log.WithField("err", err).Error("Failed to list workflows")
-				}
-				wfs := make(map[types.UID]bool)
-				for _, wf := range list {
-					wfs[wf.UID] = true
+				// get every lives workflow (1000s) into a map
+				liveOffloadNodeStatusVersions := make(map[types.UID]string)
+				for cont := ""; ; {
```

Comment:
Still, I feel we are doing the costly operation here. list all workflows and iterate them. I am trying to see we can optimize list operations to get with UID. So We just need to make very few calls based on `ListOldOffloads`. I know this PR is a show stopper for v2.5 release. I would like @jessesuen to take a quick look and suggest a better way.